### PR TITLE
Avoid false-positive warnings in Gcc 12.1.1

### LINKED ISSUE #1116: Compilation errors with Gcc 12.1.1
Trying to build Seastar with Gcc 12.1.1 (on Fedora 36), I get the following compilation errors:

1. In `json/formatter.cc`, `core/metrics.cc`, `core/scollectd.cc`, `tests/unit/sstring_test.cc`, `apps/memcached/memcache.cc` all have something calling `seastar::internal::to_sstring_sprintf(T, const char*) [with string_type = seastar::basic_sstring<char, unsigned int, 15>; T = float]’` which is `reading between 15 and 2147483647 bytes from a region of size 14 [-Werror=stringop-overread]`. Size 14 comes from `char tmp[sizeof(value) * 3 + 2]`.
2. In `tests/unit/futures_test.cc` it complains about several promises used uninitialized.

Note that Clang on Fedora 36 compiles Seastar fine - these problems are specifically with Gcc.

### PR DESCRIPTION
**Pull request by nyh**
Gcc 12.1.1 discovers what it thinks are two bugs in Seastar and warns about them  - breaking the build.
But in fact both cases are deliberate choices which we made in our implementation, and are correct. 
So we need to use "pragma"s to tell Gcc that these are in fact correct.

After these two patches, Seastar can be built with Gcc on Fedora 36 (with clang, it already worked).

Fixes #1116

### CODE REVIEW

**Comment by nyh** [comment]
ping.
This patch is necessary for out-of-the-box compilation on Fedora 36 (with gcc selected by default, and the build is aborted on warnings).

**Comment by avikivity** [comment]
Did you check that continuous integration passed the series?

**Comment by nyh** [comment]
> Did you check that continuous integration passed the series?

It didn't run. I'll see if force-pushing it again will cause it :-(

File: include/seastar/core/sstring.hh [hunk 1]
```diff
@@ -692,7 +692,13 @@ string_type to_sstring_sprintf(T value, const char* fmt) {
     char tmp[sizeof(value) * 3 + 2];
     auto len = std::sprintf(tmp, fmt, value);
     using ch_type = typename string_type::value_type;
+#pragma GCC diagnostic push
```
Comment (tchaikov) [inline include/seastar/core/sstring.hh:695]: clang claims to understand GCC's pragma with its additions. but it seems some older clang is not able to keep up with a more recent GCC. probably we have to conditionalize this on the clang version.

**Comment by nyh** [comment]
Unfortunately, the pragma I added fails on clang:
```
/home/circleci/project/include/seastar/core/sstring.hh:699:32: error: unknown warning group '-Wstringop-overread', ignored [-Werror,-Wunknown-warning-option]
#pragma GCC diagnostic ignored "-Wstringop-overread"
```

This is ironic, because the `#pragma` is explicitly labeled GCC. Which I guess is fine for clang to try to emulate, but complaining about a feature that it can't emulate as an error in the code, that's a bit over the top. I'll see how I can do something that will work on both compilers.

File: include/seastar/core/sstring.hh (hunk 1 above)
Comment (nyh) [reply include/seastar/core/sstring.hh:695]: I tested this on clang 14, so it's not old, but perhaps not new enough.
I found an even simpler workaround - add `#pragma GCC diagnostic ignored "-Wpragmas"`.
Which means, if the pragma worked, it worked. If it's not, then it's just ignored. In any case we care about the compilation working, not the pragram working :-)
I have a version I tested on both gcc and clang, I'll push it in a couple of minutes.

**Comment by nyh** [comment]
Pushed a new version which uses:
```c++
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wpragmas"
+#pragma GCC diagnostic ignored "-Wstringop-overread"
     return string_type(reinterpret_cast<ch_type*>(tmp), len);
+#pragma GCC diagnostic pop
 }
```
If the `#pragma GCC diagnostic ignored "-Wstringop-overread"` works (on gcc 12), good. If it doesn't, the "`-Wpragmas`" should ignore that it doesn't work (at least, it works on clang 14). Let's see what the CI says...

**Comment by nyh** [comment]
Wow, it doesn't work because on old clang the -Wpragmas is not enough :-(
Let's try a three-pronged approach...

**Comment by nyh** [comment]
The version that finally worked:
```c++
#pragma GCC diagnostic ignored "-Wpragmas"
#pragma GCC diagnostic ignored "-Wunknown-warning-option"
#pragma GCC diagnostic ignored "-Wstringop-overread"
```
The third line is needed in gcc 12. The second line tells clang 14 to ignore the the third line it doesn't recognize. The first line tells clang 12 to ignore the second line which it doesn't recognize :-) I don't know how to make it shorter, but it works, and it's not really that bad.

**Comment by nyh** [comment]
The two failed debug tests are because of a known and pre-existing TLS test bug. Unrelated to my patch.

**Pull request merged**